// @loaders.gl/loader-utils/dist/lib/files/http-file.js
export class HttpFile {
  constructor(url) {
    this.handle = void 0;
    this.size = 0;
    this.bigsize = 0n;
    this.url = void 0;
    this.handle = url;
    this.url = url;
  }
  async close() {}
  async stat() {
    const response = await fetch(this.handle, {
      method: 'HEAD'
    });
    if (!response.ok) {
      throw new Error(`Failed to fetch HEAD ${this.handle}`);
    }
    const size = parseInt(response.headers.get('Content-Length') || '0');
    return {
      size,
      bigsize: BigInt(size),
      isDirectory: false
    };
  }
  async read(offset, length) {
    const response = await this.fetchRange(offset, length);
    const arrayBuffer = await response.arrayBuffer();
    return arrayBuffer;
  }
  async fetchRange(offset, length, signal) {
    const nOffset = Number(offset);
    const nLength = Number(length);
    let controller;
    if (!signal) {
      controller = new AbortController();
      signal = controller.signal;
    }
    const url = this.handle;
    let response = await fetch(url, {
      signal,
      headers: {
        Range: `bytes=${nOffset}-${nOffset + nLength - 1}`
      }
    });
    switch (response.status) {
      case 206:
        break;
      case 200:
        const contentLength = response.headers.get('Content-Length');
        if (!contentLength || Number(contentLength) > length) {
          if (controller) {
            controller.abort();
          }
          throw Error('content-length header missing or exceeding request. Server must support HTTP Byte Serving.');
        }
      case 416:
        if (offset === 0) {
          const contentRange = response.headers.get('Content-Range');
          if (!contentRange || !contentRange.startsWith('bytes *')) {
            throw Error('Missing content-length on 416 response');
          }
          const actualLength = Number(contentRange.substr(8));
          response = await fetch(this.url, {
            signal,
            headers: {
              Range: `bytes=0-${actualLength - 1}`
            }
          });
        }
        break;
      default:
        if (response.status >= 300) {
          throw Error(`Bad response code: ${response.status}`);
        }
    }
    return response;
  }
}
