// @loaders.gl/loader-utils/dist/lib/sources/tile-source-adapter.js
export class TileSourceAdapter {
  constructor(source) {
    this.viewportSource = void 0;
    this.viewportSource = source;
  }
  async getMetadata() {
    return await this.viewportSource.getMetadata();
  }
  getTile(parameters) {
    const width = 512;
    const height = 512;
    const boundingBox = this.getTileBoundingBox(parameters);
    return this.viewportSource.getImage({
      boundingBox,
      layers: [],
      width,
      height
    });
  }
  getTileBoundingBox(parameters) {
    if (parameters.crs && parameters.crs !== 'ESPG3758') {
      throw new Error('SRS not ESPG3758');
    }
    const {
      x,
      y,
      zoom
    } = parameters;
    return [this.getTileLowerLeftCorner(x, y, zoom), this.getTileLowerLeftCorner(x + 1, y - 1, zoom)];
  }
  getTileLowerLeftCorner(x, y, zoom) {
    const tiles = 2 ^ zoom;
    const longitude = x / tiles * 360.0 - 180.0;
    const latitudeRadians = Math.atan(Math.sinh(Math.PI * (1 - 2 * (y + 1) / tiles)));
    const latitude = 180.0 * latitudeRadians / Math.PI;
    return [longitude, latitude];
  }
}
