// @loaders.gl/draco/dist/index.js
import { DracoLoader as DracoWorkerLoader } from "./draco-loader.js";
import DracoParser from "./lib/draco-parser.js";
import { loadDracoDecoderModule } from "./lib/draco-module-loader.js";
import { VERSION } from "./lib/utils/version.js";
export { DRACO_EXTERNAL_LIBRARIES, DRACO_EXTERNAL_LIBRARY_URLS } from "./lib/draco-module-loader.js";
export { DracoWriter } from "./draco-writer.js";
export const DracoWriterWorker = {
  id: 'draco-writer',
  name: 'Draco compressed geometry writer',
  module: 'draco',
  version: VERSION,
  worker: true,
  options: {
    draco: {},
    source: null
  }
};
export { DracoWorkerLoader };
export const DracoLoader = {
  ...DracoWorkerLoader,
  parse
};
async function parse(arrayBuffer, options) {
  const {
    draco
  } = await loadDracoDecoderModule(options);
  const dracoParser = new DracoParser(draco);
  try {
    return dracoParser.parseSync(arrayBuffer, options === null || options === void 0 ? void 0 : options.draco);
  } finally {
    dracoParser.destroy();
  }
}
