// @loaders.gl/worker-utils/dist/lib/library-utils/library-utils.js
import { isBrowser, isWorker } from "../env-utils/globals.js";
import * as node from "../node/require-utils.node.js";
import { assert } from "../env-utils/assert.js";
import { VERSION } from "../env-utils/version.js";
const loadLibraryPromises = {};
export async function loadLibrary(libraryUrl) {
  let moduleName = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : null;
  let options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  let libraryName = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;
  if (moduleName) {
    libraryUrl = getLibraryUrl(libraryUrl, moduleName, options, libraryName);
  }
  loadLibraryPromises[libraryUrl] = loadLibraryPromises[libraryUrl] || loadLibraryFromFile(libraryUrl);
  return await loadLibraryPromises[libraryUrl];
}
export function getLibraryUrl(library, moduleName) {
  let options = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : {};
  let libraryName = arguments.length > 3 && arguments[3] !== undefined ? arguments[3] : null;
  if (!options.useLocalLibraries && library.startsWith('http')) {
    return library;
  }
  libraryName = libraryName || library;
  const modules = options.modules || {};
  if (modules[libraryName]) {
    return modules[libraryName];
  }
  if (!isBrowser) {
    return `modules/${moduleName}/dist/libs/${libraryName}`;
  }
  if (options.CDN) {
    assert(options.CDN.startsWith('http'));
    return `${options.CDN}/${moduleName}@${VERSION}/dist/libs/${libraryName}`;
  }
  if (isWorker) {
    return `../src/libs/${libraryName}`;
  }
  return `modules/${moduleName}/src/libs/${libraryName}`;
}
async function loadLibraryFromFile(libraryUrl) {
  if (libraryUrl.endsWith('wasm')) {
    return await loadAsArrayBuffer(libraryUrl);
  }
  if (!isBrowser) {
    try {
      return node && node.requireFromFile && (await node.requireFromFile(libraryUrl));
    } catch (error) {
      console.error(error);
      return null;
    }
  }
  if (isWorker) {
    return importScripts(libraryUrl);
  }
  const scriptSource = await loadAsText(libraryUrl);
  return loadLibraryFromString(scriptSource, libraryUrl);
}
function loadLibraryFromString(scriptSource, id) {
  if (!isBrowser) {
    return node.requireFromString && node.requireFromString(scriptSource, id);
  }
  if (isWorker) {
    eval.call(globalThis, scriptSource);
    return null;
  }
  const script = document.createElement('script');
  script.id = id;
  try {
    script.appendChild(document.createTextNode(scriptSource));
  } catch (e) {
    script.text = scriptSource;
  }
  document.body.appendChild(script);
  return null;
}
async function loadAsArrayBuffer(url) {
  if (isBrowser || !node.readFileAsArrayBuffer || url.startsWith('http')) {
    const response = await fetch(url);
    return await response.arrayBuffer();
  }
  return await node.readFileAsArrayBuffer(url);
}
async function loadAsText(url) {
  if (isBrowser || !node.readFileAsText || url.startsWith('http')) {
    const response = await fetch(url);
    return await response.text();
  }
  return await node.readFileAsText(url);
}
