// @loaders.gl/loader-utils/dist/lib/filesystems/node-filesystem-facade.js
import { isBrowser } from "../env-utils/globals.js";
const NOT_IMPLEMENTED = new Error('Not implemented');
export class NodeFileSystemFacade {
  constructor(options) {
    var _globalThis$loaders;
    this.readable = true;
    this.writable = true;
    if ((_globalThis$loaders = globalThis.loaders) !== null && _globalThis$loaders !== void 0 && _globalThis$loaders.NodeFileSystem) {
      return new globalThis.loaders.NodeFileSystem(options);
    }
    if (isBrowser) {
      throw new Error('Can\'t instantiate NodeFileSystem in browser.');
    }
    throw new Error('Can\'t instantiate NodeFileSystem. Make sure to import @loaders.gl/polyfills first.');
  }
  async openReadableFile(path, flags) {
    throw NOT_IMPLEMENTED;
  }
  async openWritableFile(path, flags, mode) {
    throw NOT_IMPLEMENTED;
  }
  async readdir() {
    let dirname = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : '.';
    let options = arguments.length > 1 ? arguments[1] : undefined;
    throw NOT_IMPLEMENTED;
  }
  async stat(path, options) {
    throw NOT_IMPLEMENTED;
  }
  async unlink(path) {
    throw NOT_IMPLEMENTED;
  }
  async fetch(path, options) {
    throw NOT_IMPLEMENTED;
  }
}
