// @loaders.gl/worker-utils/dist/lib/worker-farm/worker-body.js
import { getTransferList } from "../worker-utils/get-transfer-list.js";
import { parentPort } from "../node/worker_threads.js";
async function getParentPort() {
  return parentPort;
}
const onMessageWrapperMap = new Map();
export default class WorkerBody {
  static async inWorkerThread() {
    return typeof self !== 'undefined' || Boolean(await getParentPort());
  }
  static set onmessage(onMessage) {
    async function handleMessage(message) {
      const parentPort = await getParentPort();
      const {
        type,
        payload
      } = parentPort ? message : message.data;
      onMessage(type, payload);
    }
    getParentPort().then(parentPort => {
      if (parentPort) {
        parentPort.on('message', handleMessage);
        parentPort.on('exit', () => console.debug('Node worker closing'));
      } else {
        globalThis.onmessage = handleMessage;
      }
    });
  }
  static async addEventListener(onMessage) {
    let onMessageWrapper = onMessageWrapperMap.get(onMessage);
    if (!onMessageWrapper) {
      onMessageWrapper = async message => {
        if (!isKnownMessage(message)) {
          return;
        }
        const parentPort = await getParentPort();
        const {
          type,
          payload
        } = parentPort ? message : message.data;
        onMessage(type, payload);
      };
    }
    const parentPort = await getParentPort();
    if (parentPort) {
      console.error('not implemented');
    } else {
      globalThis.addEventListener('message', onMessageWrapper);
    }
  }
  static async removeEventListener(onMessage) {
    const onMessageWrapper = onMessageWrapperMap.get(onMessage);
    onMessageWrapperMap.delete(onMessage);
    const parentPort = await getParentPort();
    if (parentPort) {
      console.error('not implemented');
    } else {
      globalThis.removeEventListener('message', onMessageWrapper);
    }
  }
  static async postMessage(type, payload) {
    const data = {
      source: 'loaders.gl',
      type,
      payload
    };
    const transferList = getTransferList(payload);
    const parentPort = await getParentPort();
    if (parentPort) {
      parentPort.postMessage(data, transferList);
    } else {
      globalThis.postMessage(data, transferList);
    }
  }
}
function isKnownMessage(message) {
  const {
    type,
    data
  } = message;
  return type === 'message' && data && typeof data.source === 'string' && data.source.startsWith('loaders.gl');
}
