// @loaders.gl/loader-utils/dist/lib/file-provider/file-handle-file.js
import { NodeFileFacade as NodeFile } from "../files/node-file-facade.js";
export class FileHandleFile {
  constructor(path) {
    this.file = void 0;
    this.size = void 0;
    this.file = new NodeFile(path, 'r');
    this.size = this.file.bigsize;
  }
  async destroy() {
    await this.file.close();
  }
  async getUint8(offset) {
    const arrayBuffer = await this.file.read(offset, 1);
    const val = new Uint8Array(arrayBuffer).at(0);
    if (val === undefined) {
      throw new Error('something went wrong');
    }
    return val;
  }
  async getUint16(offset) {
    const arrayBuffer = await this.file.read(offset, 2);
    const val = new Uint16Array(arrayBuffer).at(0);
    if (val === undefined) {
      throw new Error('something went wrong');
    }
    return val;
  }
  async getUint32(offset) {
    const arrayBuffer = await this.file.read(offset, 4);
    const val = new Uint32Array(arrayBuffer).at(0);
    if (val === undefined) {
      throw new Error('something went wrong');
    }
    return val;
  }
  async getBigUint64(offset) {
    const arrayBuffer = await this.file.read(offset, 8);
    const val = new BigInt64Array(arrayBuffer).at(0);
    if (val === undefined) {
      throw new Error('something went wrong');
    }
    return val;
  }
  async slice(startOffset, endOffset) {
    const bigLength = endOffset - startOffset;
    if (bigLength > Number.MAX_SAFE_INTEGER) {
      throw new Error('too big slice');
    }
    const length = Number(bigLength);
    return await this.file.read(startOffset, length);
  }
  get length() {
    return this.size;
  }
}
