// @loaders.gl/schema/dist/lib/table/arrow-api/arrow-like-table.js
import { ArrowLikeSchema } from "./arrow-like-schema.js";
import { deduceTableSchema } from "../simple-table/table-schema.js";
import { getTableCell, getTableLength, getTableNumCols } from "../simple-table/table-accessors.js";
class ArrowLikeVector {
  constructor(table, columnName) {
    this.table = void 0;
    this.columnName = void 0;
    this.table = table;
    this.columnName = columnName;
  }
  get(rowIndex) {
    return getTableCell(this.table, rowIndex, this.columnName);
  }
  toArray() {
    var _arrowTable$getChild;
    switch (this.table.shape) {
      case 'arrow-table':
        const arrowTable = this.table.data;
        return (_arrowTable$getChild = arrowTable.getChild(this.columnName)) === null || _arrowTable$getChild === void 0 ? void 0 : _arrowTable$getChild.toArray();
      case 'columnar-table':
        return this.table.data[this.columnName];
      default:
        throw new Error(this.table.shape);
    }
  }
}
export class ArrowLikeTable {
  constructor(table) {
    this.schema = void 0;
    this.table = void 0;
    const schema = table.schema || deduceTableSchema(table);
    this.schema = new ArrowLikeSchema(schema.fields, schema.metadata);
    this.table = {
      ...table,
      schema
    };
  }
  get data() {
    return this.table.shape === 'geojson-table' ? this.table.features : this.table.data;
  }
  get numCols() {
    return getTableNumCols(this.table);
  }
  get length() {
    return getTableLength(this.table);
  }
  getChild(columnName) {
    return new ArrowLikeVector(this.table, columnName);
  }
}
